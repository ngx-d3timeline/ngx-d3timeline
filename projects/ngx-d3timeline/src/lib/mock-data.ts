import { TimelineEvent } from './timeline-event';

export const mockData: TimelineEvent[] = [
  {
    id: 1,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 2,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 3,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 1',
    sequence: 1
  },
  {
    id: 13,
    start: new Date('2017-09-22T10:00:00'),
    finish: new Date('2017-09-22T10:08:31'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 14,
    start: new Date('2017-09-22T09:15:00'),
    finish: new Date('2017-09-22T09:16:16'),
    type: 'Waiting',
    series: 'Series 2',
    sequence: 6
  },
  {
    id: 15,
    start: new Date('2017-09-22T13:45:00'),
    finish: new Date('2017-09-22T14:52:27'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 41,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 42,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 43,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 3',
    sequence: 1
  }
];
