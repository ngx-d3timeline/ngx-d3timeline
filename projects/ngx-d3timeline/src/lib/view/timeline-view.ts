export class TimelineView {
  readonly width: number;
  readonly height: number;
  readonly margin: number;

  readonly bottom: number;
  readonly left: number;
  readonly right: number;
  readonly top: number;

  constructor({
    width,
    height,
    margin
  }: {
    width: number;
    height: number;
    margin: number;
  }) {
    this.width = width;
    this.height = height;
    this.margin = margin;

    this.bottom = height - margin;
    this.left = 0;
    this.right = width - margin;
    this.top = 0;
  }
}
