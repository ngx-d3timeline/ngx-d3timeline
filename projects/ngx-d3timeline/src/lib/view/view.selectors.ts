import { selectView } from '../store/state';
import {
  getViewTopLeft,
  getViewClipRectangle,
  getViewClipRectangleHeight,
  getViewClipRectangleWidth
} from './view-utils';
import { createSelector } from '../store-lib/selector/create-selector';

export const selectViewTopLeft = createSelector(selectView, getViewTopLeft);

export const selectViewClipRectangleHeight = createSelector(
  selectView,
  getViewClipRectangleHeight
);

export const selectViewClipRectangleWidth = createSelector(
  selectView,
  getViewClipRectangleWidth
);

export const selectViewClipRectangle = createSelector(
  selectViewTopLeft,
  selectViewClipRectangleWidth,
  selectViewClipRectangleHeight,
  getViewClipRectangle
);
