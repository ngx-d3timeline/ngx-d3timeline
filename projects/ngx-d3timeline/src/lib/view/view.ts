export class View {
  readonly width: number;
  readonly height: number;
  readonly margin = 50;

  readonly left = this.margin;
  readonly top = this.margin;
  readonly right: number;
  readonly bottom: number;

  constructor([width, height]: [number, number]) {
    this.width = width;
    this.height = height;

    this.bottom = height - this.margin;
    this.right = width - this.margin;
  }
}
