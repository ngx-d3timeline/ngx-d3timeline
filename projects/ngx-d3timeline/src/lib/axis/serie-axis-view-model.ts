import { ScaleBand, scaleBand } from 'd3-scale';
import { TimelineEvent } from '../timeline-event';
import { View } from '../view/view';

export class SeriesAxisViewModel {
  private readonly bandscale: ScaleBand<string>;

  constructor(data: TimelineEvent[], view: View) {
    this.bandscale = this.createBandscale(data, view);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.bandscale.domain().map(tick => ({
      label: tick,
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.bandscale.range()[1];
  }

  get bandwidth() {
    return this.bandscale.bandwidth();
  }

  private createBandscale(data: TimelineEvent[], view: View) {
    return scaleBand()
      .domain([...new Set(data.map(d => d.series))])
      .range([0, view.width - view.margin]);
  }

  private tickTransform(tick: string) {
    return `translate(${this.bandscale(tick) +
      this.bandscale.bandwidth() / 2}, 0)`;
  }
}
