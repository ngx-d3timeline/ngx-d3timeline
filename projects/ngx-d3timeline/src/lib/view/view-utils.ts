import { View } from './view';
import { Point } from '../core/point';
import { Rectangle } from '../core/rectangle';

export function getViewTopLeft(view: View): Point {
  return { x: view.left, y: view.top };
}

export function getViewClipRectangle(view: View): Rectangle {
  return {
    width: view.width - 2 * view.margin,
    height: view.height - 2 * view.margin,
    x: view.left,
    y: view.top
  };
}
