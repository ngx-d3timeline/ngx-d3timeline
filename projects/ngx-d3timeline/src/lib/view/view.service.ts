import { Injectable } from '@angular/core';
import { TimelineView } from './timeline-view';
import { BehaviorSubject } from 'rxjs';
import { map, tap } from 'rxjs/operators';

@Injectable({ providedIn: 'root' })
export class ViewService {
  readonly margin = 50;

  private viewSubject = new BehaviorSubject<[number, number]>(null);
  view$ = this.viewSubject.pipe(
    map(([width, height]) => new TimelineView(width, height, this.margin))
  );

  get rootTransform() {
    return `translate(${this.margin}, ${this.margin})`;
  }

  setView([width, height]: [number, number]) {
    this.viewSubject.next([width, height]);
  }
}
