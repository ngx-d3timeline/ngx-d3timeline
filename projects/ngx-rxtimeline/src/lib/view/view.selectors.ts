import { selectView } from '../store/state';

import { createSelector } from '../store-lib/selector/create-selector';
import { Point } from '../core/point';
import {
  createStructuredSelector,
  createEnumSelector
} from '../store-lib/selector/selector-utils';
import { Rectangle } from '../core/rectangle';
import { toArray, sum } from '../core/array-utils';
import { pipe, clampZero, subtract, double } from '../core/function-utils';
import { constSelector } from '../store-lib/selector/selector';
import { Orientation } from '../core/orientation';
import {
  selectTimeOrientation,
  selectAxisOrientation
} from '../options/selectors/options.selectors';
import { AxisType } from '../axis/axis';

export const selectVerticalMargin = constSelector(50);
export const selectMarginRight = constSelector(50);
const selectMarginLeftWhenHorizontal = constSelector(100);
const selectMarginLeftWhenVertical = constSelector(50);

export const selectMarginLeft = createEnumSelector<Orientation, number>({
  Horizontal: selectMarginLeftWhenHorizontal,
  Vertical: selectMarginLeftWhenVertical
})(selectTimeOrientation);

const selectVerticalMarginTotal = createSelector(selectVerticalMargin, double);
const selectHorizontalMarginTotal = createSelector(
  selectMarginLeft,
  selectMarginRight,
  sum
);

const selectViewWidth = createSelector(selectView, view => view && view.width);
const selectViewHeight = createSelector(
  selectView,
  view => view && view.height
);

const selectViewLeft = createSelector(
  selectViewWidth,
  selectMarginLeft,
  Math.min
);

const selectViewTop = createSelector(
  selectViewHeight,
  selectVerticalMargin,
  Math.min
);

const selectViewRight = createSelector(
  selectViewWidth,
  selectMarginRight,
  subtract
);

const selectViewBottom = createSelector(
  selectViewHeight,
  selectVerticalMargin,
  subtract
);

export const selectViewTopLeft = createStructuredSelector<Point>({
  x: selectViewLeft,
  y: selectViewTop
});

export const selectViewClipRectHeight = createSelector(
  selectViewHeight,
  selectVerticalMarginTotal,
  pipe(subtract, clampZero)
);

export const selectViewClipRectWidth = createSelector(
  selectViewWidth,
  selectHorizontalMarginTotal,
  pipe(subtract, clampZero)
);

export const selectViewClipRect = createStructuredSelector<Rectangle>({
  height: selectViewClipRectHeight,
  width: selectViewClipRectWidth,
  x: selectViewLeft,
  y: selectViewTop
});

export const selectViewVerticalRange = createSelector(
  selectViewTop,
  selectViewBottom,
  toArray
);

export const selectViewHorizontalRange = createSelector(
  selectViewLeft,
  selectViewRight,
  toArray
);

export const selectViewRange = (axisType: AxisType) =>
  createEnumSelector<Orientation, number[]>({
    Vertical: selectViewVerticalRange,
    Horizontal: selectViewHorizontalRange
  })(selectAxisOrientation(axisType));
