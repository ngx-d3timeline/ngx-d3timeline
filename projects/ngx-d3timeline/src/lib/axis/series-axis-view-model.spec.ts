import { SeriesAxisViewModel } from './serie-axis-view-model';
import { mockData } from '../mock-data';
import { TimelineView } from '../view/timeline-view';

describe('SeriesAxisViewModel', () => {
  let vm: SeriesAxisViewModel;
  const timelineView: TimelineView = new TimelineView(300, 400, 50);

  beforeEach(() => {
    vm = new SeriesAxisViewModel(mockData, timelineView);
  });

  it('should return correct tickInfo', () => {
    expect(vm.tickInfos).toMatchSnapshot();
  });

  it('should return correct rangeLimit', () => {
    expect(vm.rangeLimit).toBe(timelineView.bounds.right);
  });
});
