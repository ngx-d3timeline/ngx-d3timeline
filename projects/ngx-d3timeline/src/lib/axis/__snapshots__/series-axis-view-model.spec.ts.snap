// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesAxisViewModel should return correct tickInfo 1`] = `
Array [
  Object {
    "label": "Series 1",
    "transform": "translate(41.666666666666664, 0)",
  },
  Object {
    "label": "Series 2",
    "transform": "translate(125, 0)",
  },
  Object {
    "label": "Series 3",
    "transform": "translate(208.33333333333331, 0)",
  },
]
`;
